### ISSUE TITLE
A rule suggesting to replace `Object.prototype.hasOwnProperty.call` with `Object.hasOwn`

### ISSUE THREAD

**Issue by pubmikeb** [eslint/eslint#14939]
**Please describe what the rule should do:**
Starting V8 v.9.3, `Object.prototype.hasOwnProperty.call` can be replaced with an alias/syntax sugar `Object.hasOwn`, which is much more read-friendly. Further information: https://v8.dev/features/object-has-own

**What new ECMAScript feature does this rule relate to?**
Promoting using of an alias/syntax sugar `Object.hasOwn` instead of `Object.prototype.hasOwnProperty.call`.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[X] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**
The original code:
```js
if (Object.prototype.hasOwnProperty.call(object, 'foo')) {
  // `object` has property `foo`.
}
```

Should be replaced with:
```js
if (Object.hasOwn(object, 'foo')) {
  // `object` has property `foo`.
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
Applying this rule you can make your application's code cleaner.

**Are you willing to submit a pull request to implement this rule?**
No.

**Comment by nzakas** [comment]
I will champion this rule, as I think it will help people better understand this new method.

**Comment by bmish** [comment]
FYI this rule was implemented in another project as [unicorn/prefer-object-has-own](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-object-has-own.md) but I suspect @fisker would be happy to see it moved to ESLint core where it can benefit more people.

**Comment by nzakas** [comment]
@bmish oh cool! We do typically introduce new rules to help people transition from old ways of doing things to new ways of doing things, so we might still end up with it as a core rule. Let’s get a few more thoughts on this. (Especially since there isn’t a great way to discover such rules and plug-ins.)

**Comment by bmish** [comment]
Yep, I'm also in favor of new core rules for that reason (like this one).

**Comment by fisker** [comment]
Will be happy to see the rule in core rules.

**Comment by aladdin-add** [comment]
[proposal-accessible-object-hasownproperty](https://github.com/tc39/proposal-accessible-object-hasownproperty) is at stage-3.  we can reconsider it when it reaches stage-4.

**Comment by pubmikeb** [comment]
> [proposal-accessible-object-hasownproperty](https://github.com/tc39/proposal-accessible-object-hasownproperty) is at stage-3. we can reconsider it when it reaches stage-4.

Since this feature is already implemented in V8 and will be in GA in Chromium-based browsers/Firefox/Safari/Node.js by the end of August, the reaching stage-4 is just a formality.

**Comment by nzakas** [comment]
@aladdin-add we wait for stage 4 for new syntax, but new APIs are okay to address in stage 3.

**Comment by mdjermanovic** [comment]
I'm in favor of this rule, and I think it should also report `{}.hasOwnProperty.call(object, 'foo')`.

**Comment by ZhangChengLin** [comment]
> I'm in favor of this rule, and I think it should also report `{}.hasOwnProperty.call(object, 'foo')`.


```
function fun(){
    return arguments.length
}
```
How to write arguments so as not to be prompted no-undef

**Comment by btmills** [comment]
I’m also supportive. Marking as accepted!

**Comment by thisiskartikgupta** [comment]
Hello everyone, I am a beginner to open-source. I have some experience in working with Javascript. Can I work on this issue?

**Comment by snitin315** [comment]
@thisiskartikgupta sure, let us know if you need any help.

**Comment by jamiebuilds** [comment]
FYI: `Object.hasOwn()` just reached Stage 4

**Comment by eyeamkd** [comment]
Is the PR as simple as just replacing the code? Or is there any other problem to consider?

**Comment by nzakas** [comment]
The rule first needs to identify the patterns mentioned in the issue. The second step is to replace. It is fairly straightforward. 

@thisiskartikgupta are you still working on this?

**Comment by eyeamkd** [comment]
Understood @nzakas. If @thisiskartikgupta  isn't currently working on this, can you please assign this to me?

**Comment by thisiskartikgupta** [comment]
I'm not working on it. You're good to go @eyeamkd !!

**Comment by nzakas** [comment]
@eyeamkd it’s all yours.

**Comment by eyeamkd** [comment]
![Screenshot 2021-09-14 at 7 17 30 AM](https://user-images.githubusercontent.com/31094327/133181453-62a79f7c-30fc-463a-b204-07927b194e14.png)

doing npm install gives me this, is it okay to do npm i -f and continue? prior apologies if this is a trivial question

**Comment by btmills** [comment]
@eyeamkd yes, `npm install --force` will do the trick until we release v8.0.0 final

**Comment by eyeamkd** [comment]
Worked @btmills

**Comment by bmish** [comment]
Just a reminder that there's an existing rule [unicorn/prefer-object-has-own](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-object-has-own.md) that we should be able to copy (and it has been battle-tested for several months).

**Comment by eyeamkd** [comment]
Alright, so all I gotta do is just take the rule and apply conditions on to it? @bmish

**Comment by eyeamkd** [comment]
Okay!

**Comment by nzakas** [comment]
Hold on! We aren’t just going to copy someone else’s work and paste it into ESLint. We should create our own to avoid licensing issues.

This is a fairly straightforward rule that shouldn’t require anything complex enough that is required copying something directly.

**Comment by nzakas** [comment]
@eyeamkd if you need a starting point, you can copy this rule:

https://eslint.org/docs/rules/prefer-object-spread

The logic for finding Object.assign is exactly the same as for Object.hasOwn.

**Comment by eyeamkd** [comment]
Using the same @nzakas, and yeah won't be copy-pasting anyone else's work

**Comment by bmish** [comment]
Let me try again with better advice :) Write our own rule and then check any existing third-party rules to make sure we don't miss important test cases.

**Comment by nzakas** [comment]
@bmish 👍

**Comment by AtulKarn** [comment]
Hi! is @eyeamkd still working on this? If not then can I take this up? I am new to open source and this seemed a good issue to work on

**Comment by eyeamkd** [comment]
Yeah, will place a PR soon @AtulKarn

**Comment by AtulKarn** [comment]
Oh that's great

**Comment by nzakas** [comment]
@Gautam-Arora24 go for it!

**Comment by Gautam-Arora24** [comment]
Thanks!

**Comment by nzakas** [comment]
It looks like you don’t have an updated repo. Make sure you sync with upstream, then delete your node_modules directory and reinstall.

**Comment by bharathks005** [comment]
Is the issue is not fixed, can I start to fix this issue? can I try?

**Comment by snitin315** [comment]
It is currently being implemented in https://github.com/eslint/eslint/pull/15346.
